Sprint-3/revise/implement/password-validator.test.js

```diff
@@ -14,3 +14,92 @@ To be valid, a password must:
 
 You must breakdown this problem in order to solve it. Find one test case first and get that working
 */
+
+function isValidLength(password) {
+    if (password.length >= 5){
+        return true 
+    } else if (password.length < 5) 
+      return false;
+}
+function containsUpperCase(password) {
+    for(let i = 0 ; i < password.length; i++){
+      //console.log(password[i]);
+      if (password[i] === password[i].toUpperCase())
+         return true
+    }return false;
+}
+
+function containsLowerCase(password) {
+    for(let i = 0 ; i < password.length; i++){
+      if (password[i] === password[i].toLowerCase())
+         return true
+    }return false;
+}
+
+function hasOneAlphaNumericChar(password){
+  for(let i = 0 ; i < password.length; i++){
+    const alphaNumerics = ["!", "#", "$", "%", ".", "*", "&"]
+      if (alphaNumerics.includes(password[i]))
+         return true
+    }return false;
+}
+
+function hasOneNumber(password){
+  for(let i = 0 ; i < password.length; i++){
+    const digits = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9",]
+      if (digits.includes(password[i]))
+         return true
+    }return false;
+}
+
+function notPreviousPassword(password){
+    let previousPassword = ["Hamster2", "Dog!",]
+      if (previousPassword.includes(password)){
+         return false
+    } return true;
+  }
+
+  function isValidPassword(password){
+    if (!isValidLength(password)){
+      return false
+    }
+
+     if (!containsUpperCase(password)){
+      return false
+    }
+
+     if (!containsLowerCase(password)){
+      return false
+    }
+
+     if (!hasOneAlphaNumericChar(password)){
+      return false
+    }
+
+     if (!hasOneNumber(password)){
+      return false
+    }
+
+     if (!notPreviousPassword(password)){
+      return false
+    }
+    return true;
+  }
+
+
+
+//console.log(isValidLength("cat"));
+//console.log(containsUpperCase("cAt"));
+//console.log(containsUpperCase("cAt"));      
+//console.log(hasOneNumber("cAt"));  
+//console.log(hasOneAlphaNumericChar("cat!"))
+//console.log(notPreviousPassword("Dog!"))
+console.log(isValidPassword("!Ca5"));
+
+
+test('if password is a valid password.', () => {
+  expect(isValidPassword("abcd")).toBe(false);     
+  expect(isValidPassword("Abc!")).toBe(false);    
+  expect(isValidPassword("ABCD1")).toBe(false);    
+  expect(isValidPassword("Dog!")).toBe(false);
```

Comment:
These tests are not comprehensive, and they could not indicate exactly why the function fails. A better way to prepare the tests are in the following manner:
```
test('Check password without an uppercase letter', () => {
  expect(isValidPassword("1a!123")).toBe(false);
  expect(isValidPassword("za123!")).toBe(false);  
});

test('Check passwords without a digit', () => {
  expect(isValidPassword("!!aABC")).toBe(false);
  expect(isValidPassword("Abaaa!")).toBe(false);  
});

...
```
A comprehensive test can help you detect bugs in `isValidPassword()`.